src/core/event/events.ts

```diff
@@ -0,0 +1,86 @@
+import EventType from "./eventType";
+import Event from "./event";
+import EventState from "./eventState";
+
+type ListenerFunction = (payload: any) => void;
+
+type EventListener = {
+  callback: ListenerFunction;
+  eventState: EventState;
+};
+
+export default class Events {
+  private listeners: { [key in EventType]?: EventListener[] };
+
+  constructor() {
+    this.listeners = {};
+  }
+
+  on(
+    eventType: EventType,
+    callback: ListenerFunction,
+    eventState: EventState = EventState.POST_EVENT,
+  ): void {
+    if (!this.listeners[eventType]) {
+      this.listeners[eventType] = [];
+    }
+    this.listeners[eventType]!.push({ callback, eventState });
+  }
+
+  emit(event: Event): void {
+    const listeners = this.listeners[event.eventType];
+    if (listeners) {
+      listeners.slice().forEach((listener) => {
+        if (listener.eventState === event.eventState) {
+          listener.callback(event.payload);
```

Comment:
Just to confirm, can we ignore any concurrency-related issues here due to the nature of JS? As in that it's safe to assume `emit` will be called from the main thread in all use cases.

Reply:
Yep, I think so. One emit call will fully complete before another one begins and callback functions are synchronous. (JS takes care of the queue like execution)